keps/sig-api-machinery/20190726-component-config-fields-should-be-optional.md

```diff
@@ -0,0 +1,197 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+`ComponentConfig` allows for specification of options for a component in a
+structured and versioned way; kubernetes apimachinery objects replace flags.
+
+However, while flags make the distinction between not-specified and
+set-to-empty-value, our api types generally do not.  By using pointer-types it
+is possible to preserve this difference, this KEP proposes that we should use
+pointer-types in our componentconfig.
+
+This also allows for generation of a minimal componentconfig from code, and
+should faciliate patching and merging of componentconfig.
+
+ComponentConfig types are used with v1alpha1 versions for kube-scheduler,
+kube-proxy, kube-controller-manager, cloud-controller-manager; I intend to make
+these nilable.  kubelet uses a ComponentConfig type that is v1beta1, I propose
+not changing those at the current time, but rather proving out the approach with
+the v1alpha1 types.
+
+## Motivation
+
+I initially raised this in a [mailing list
+thread](https://groups.google.com/d/topic/kubernetes-wg-component-standard/hO7Lmi0cwQU/discussion)
+with wg-component-standard.  When trying to create componentconfig from code,
+all fields were populated, which did not really reflect the intent.  For minimal
+output, either simply for hygene or to support kustomize, we would have needed
+another type.  We would then have had to keep that type in sync with changes to
+componentconfig.  It seemed that componentconfig could better serve as that
+type.
+
+Supporting merging of configurations that better reflects intent has come up a
+few times: there have been discussions of merging fields with flags and the
+precedence when doing so, merging fields set at different levels of a hierarchy
```

Comment:
I get what you're saying, but I don't think most others would.  An example on hierarchical ordering I think would help make your case.